plugins/in_systemd/systemd.c

```diff
@@ -277,14 +283,27 @@ static int in_systemd_collect(struct flb_input_instance *i_ins,
         ctx->pending_records = FLB_FALSE;
         return FLB_SYSTEMD_OK;
     }
-
+    else if (ret_j > 0) {
     /*
-     * ret_j == -1, the loop was broken due to some special condition like
+     * ret_j == 1, but the loop was broken due to some special condition like
      * buffer size limit or it reach the max number of rows that it supposed to
      * process on this call. Assume there are pending records.
      */
     ctx->pending_records = FLB_TRUE;
     return FLB_SYSTEMD_MORE;
+    }
+    else {
+        /* Supposedly, current cursor points to a deleted file.
+         * Re-seeking to the first journal entry.
+         * Other failures, such as disk read error, would still lead to infinite loop there,
+         * but at least FLB log will be full of errors. */
+        ret = sd_journal_seek_head(ctx->j);
```

Comment:
If journal notify us that a file being read was deleted/rotated, is it safe to consider to get back to the first entry of the journal ?, do we need to update the cursor in the database if enabled ?

I am not sure about the default behavior... what do you think ?

Reply:
> If journal notify us that a file being read was deleted/rotated, is it safe to consider to get back to the first entry of the journal ?

Short answer: no.
Long answer: Systemd's [sd_journal_process() sends SD_JOURNAL_INVALIDATE](https://www.freedesktop.org/software/systemd/man/sd_journal_get_fd.html) notification when some journal file gets rotated. But that file is not necesserily the file which is being read by sd_journal_next(). I mean, journald deletes the oldest log files all the time, but FLB is already reading the most fresh log file, and thus not affected by rotation.

Also, rotation can happen during the loop inside in_systemd_collect(). Therefore SD_JOURNAL_INVALIDATE notification will be handled only on a next in_systemd_collect() run.

Therefore I cannot just re-seek the journal every time SD_JOURNAL_INVALIDATE notification arrives. This event is purely informational fur our use case. We should read the journal sequentially when possible, and just seek to the beginning if something went terribly wrong.

> , do we need to update the cursor in the database if enabled ?

I don't think so. If FLB hits this race condition over a deleted file, then in_systemd_collect() will be restarted immediately, because `ctx->pending_records = FLB_TRUE;` is set on error. Thus, new valid DB cursor will be written already on the next iteration.

And even if FLB will be killed between detecting a journal read error (=re-seek) and re-run of in_systemd_collect(), the newly restarted instance will read the last processed journal pointer from sqlite. FLB will try to seek to that (invalid) pointer, will be unable to seek to a point in a deleted file, and [will seek implicitly to the first available entry](https://www.freedesktop.org/software/systemd/man/sd_journal_seek_cursor.html):
>> _If no entry matching the specified cursor is found the call will seek to the next closest entry (in terms of time) instead._

> I am not sure about the default behavior... what do you think ?

I think it is save to continue with the way it is implemented now.
We run hundreds of FLB instances using this code without issue so far.